src/sqlfluff/rules/L067.py

```diff
@@ -0,0 +1,160 @@
+"""Implementation of Rule L067."""
+
+from sqlfluff.core.parser import NewlineSegment
+from sqlfluff.core.rules.base import BaseRule, LintFix, LintResult, RuleContext
+from sqlfluff.core.rules.doc_decorators import (
+    document_configuration,
+    document_fix_compatible,
+    document_groups,
+)
+import sqlfluff.core.rules.functional.segment_predicates as sp
+
+
+@document_groups
+@document_configuration
+@document_fix_compatible
+class Rule_L067(BaseRule):
+
+    """A subquery inside parentheses should go a new line.
+        For brackets, if there are any subquery inside, the subqueries
+        should take the next line after opening, and the ending bracket
+        should take the next line after the subqueries.
+
+    **Anti-pattern**
+
+    Parentheses on the line with a subquery.
+
+    .. code-block:: sql
+        :force:
+
+        SELECT A.SOMETHING
+        FROM (SELECT
+                B,
+                C
+            FROM D GROUP BY 1) AS A
+
+
+
+    **Best practice**
+
+    Opening bracket and ending bracket take a new line out of the subquery.
+
+    .. code-block:: sql
+        :force:
+
+        SELECT A.SOMETHING
+        FROM (
+            SELECT
+                B,
+                C
+            FROM D GROUP BY 1
+        ) AS A
+
+    """
+    groups = ("all",)
+
+    def _eval(self, context: RuleContext):
+        if context.segment.is_type('bracketed'):
+            expressions = context.functional.segment.children()
+            start_bracket = context.segment.start_bracket[0] # Tuple of length 1
+            end_bracket = context.segment.end_bracket[0]
+
+            # Check whether there is any subquery found inside the brackets
+            select_targets = expressions.select(sp.is_type("select_statement"))
+            if select_targets:
+                results = []
+                # check if the select statement only take a single line.
+                # If so, the parentheses can keep at the same line with them.
+                if (
+                    len(select_targets) == 1
+                    and select_targets[0].get_start_loc()[0]
+                    == select_targets[0].get_end_loc()[0]
+                    == start_bracket.pos_marker.line_no
+                    == end_bracket.pos_marker.line_no
```

Comment:
If I recall correctly, location and position information is not updated during each "loop" of the linting operation, so this may not behave as expected, e.g. if an earlier lint rule has already made changes. (L016 is a likely example, since it splits long lines.)

Can you try testing this? First manually, but we should probably add an automated test for it. Tests that involve multiple rules are usually defined in `test/fixtures/linter/autofix/ansi`.

The more robust way to check for positioning is to examine the actual segments and their contents, which are always "current".

Reply:
It seems that L016 or some other function could still update the locations, but not to all segments.
See this interesting example:
```
SELECT A.SOMETHING
FROM (SELECT ThisIsALongExampleOfColumnInTable FROM ThisIsAVeryLongTableNameInOrderToTest GROUP BY 1) AS A
```
In the loop, 
select_targets[0].pos_marker.line_no is 2, 
select_targets[0].get_start_loc() return (3, 6)
But brackets' locations are not updated.

It's true that the position information is accurate during the loop.
What do you mean "examine the actual segments and their contents" ? Is it good to check whether there is any newline inner the bracket?

Reply:
I think you need to iterate the parse tree to look for segments of type `newline`. Rule L004 also uses a similar technique.